File: 프로그래머스/0/120840. 구슬을 나누는 경우의 수/구슬을 나누는 경우의 수.swift
import Foundation

func solution(_ balls:Int, _ share:Int) -> Int {
    if share == 0 || balls == share { return 1 } // 0개를 선택하거나 모두 선택할 경우
    let r = balls - share
    let m = share
    let smaller = min(r, m) // 더 작은 값을 사용하여 범위 최적화
    let larger = max(r, m)
    
    let numerator = (larger+1...balls).reduce(1.0) { $0 * Double($1) }
    let denominator = (1...smaller).reduce(1.0) { $0 * Double($1) }
    
    return Int(numerator / denominator)
}

File: 프로그래머스/0/120840. 구슬을 나누는 경우의 수/구슬을 나누는 경우의 수.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

int solution(int balls, int share) {
    int answer = 0;
    __int128 x = 1, y = 1;

    if (balls - share < share) {
        share = balls - share;
    }

    for (int i = balls - share + 1; i <= balls; i++) {
        x *= i;
    }

    for (int i = 1; i <= share; i++) {
        y *= i;
    }

    answer = x / y;

    return answer;
}